src/io.c

```diff
@@ -72,76 +66,40 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
     return len + 1;
 }
 
-uint32_t memory_ifetch(memory_t *m, uint32_t addr)
-{
-    const uint32_t addr_lo = addr & mask_lo;
-    assert((addr_lo & 1) == 0);
-
-    chunk_t *c = m->chunks[addr >> 16];
-    assert(c);
-    return *(const uint32_t *) (c->data + addr_lo);
-}
-
-uint32_t memory_read_w(memory_t *m, uint32_t addr)
+uint32_t memory_ifetch(uint32_t addr)
```

Comment:
Let's inline short functions like `memory_ifetch` in header files.

Reply:
we cannot define it as inline function, because we assign it as function pointer to struct `riscv_io_t`.

Reply:
Confirmed, the compiler cannot inline a function unless it can determine that only one possible version of the function will be called. When calling through a pointer, it is not trivially obvious that this is the case.

However, we may still look for the approaches to lower function invocation cost later.

Reply:
Append TODO in git commit messages for the possibility to inline the essential memory operations.